### PR TITLE
Stabilize `atomic_from_ptr`

### PR DESCRIPTION
**Pull request by tgross35**
This stabilizes `atomic_from_ptr` and moves the const gate to `const_atomic_from_ptr`. Const stability is blocked on `const_mut_refs`.

Tracking issue:  #108652

Newly stable API:

```rust
// core::atomic

impl AtomicBool { pub unsafe fn from_ptr<'a>(ptr: *mut bool) -> &'a AtomicBool; }

impl<T> AtomicPtr<T> { pub unsafe fn from_ptr<'a>(ptr: *mut *mut T) -> &'a AtomicPtr<T>; }

impl AtomicU8    { pub unsafe fn from_ptr<'a>(ptr: *mut u8)    -> &'a AtomicU8;    }
impl AtomicU16   { pub unsafe fn from_ptr<'a>(ptr: *mut u16)   -> &'a AtomicU16;   }
impl AtomicU32   { pub unsafe fn from_ptr<'a>(ptr: *mut u32)   -> &'a AtomicU32;   }
impl AtomicU64   { pub unsafe fn from_ptr<'a>(ptr: *mut u64)   -> &'a AtomicU64;   }
impl AtomicUsize { pub unsafe fn from_ptr<'a>(ptr: *mut usize) -> &'a AtomicUsize; }

impl AtomicI8    { pub unsafe fn from_ptr<'a>(ptr: *mut i8)    -> &'a AtomicI8;    }
impl AtomicI16   { pub unsafe fn from_ptr<'a>(ptr: *mut i16)   -> &'a AtomicI16;   }
impl AtomicI32   { pub unsafe fn from_ptr<'a>(ptr: *mut i32)   -> &'a AtomicI32;   }
impl AtomicI64   { pub unsafe fn from_ptr<'a>(ptr: *mut i64)   -> &'a AtomicI64;   }
impl AtomicIsize { pub unsafe fn from_ptr<'a>(ptr: *mut isize) -> &'a AtomicIsize; }
```

### CODE REVIEW

**Comment by tgross35** [comment]
This seems like a pretty uncontroversial method that mirrors `AtomicX::as_ptr`. Requres FCP

r? libs-api
@rustbot label +t-libs-api -t-libs

**Review by dtolnay — APPROVED** [review]

**Comment by dtolnay** [comment]
@rust-lang/libs-api:
@rfcbot fcp merge

**Comment by WaffleLapkin** [comment]
Hm, looking at the docs again,

> * The value behind `ptr` must not be accessed through non-atomic operations for the whole lifetime `'a`.

This doesn't seem correct to me (or, rather, this is sufficient, but we could, and probably should, be looser). I think in practice, we only need a guarantee that the periods of using atomic and non-atomic accesses are synchronized. I.e. that for any two atomic and non-atomic accesses, there is a happens before relationship between them. (I could be wrong though, I'm not very knowledgeable here...)

**Comment by tgross35** [comment]
Do you think it would be better to say something about needing to use atomic accesses specifically whenever you cross `Sync` contexts?

I don’t think it is too bad as-is because users of the Atomic will probably assume that you can safely use it everywhere, send it across threads etc. If the caller assumes that there is a time when it doesn’t need to do atomic accesses (one thread, no race conditions) and then something down the line gets refactored to share the Atomic between threads, that could be an easy quiet race condition. Less chance for error if the caller always has to assume that it always has to use atomics.

Maybe wreaking from `must` to `should` would be enough?

**Comment by WaffleLapkin** [comment]
Well, I'm just saying that the current documentation doesn't seem to reflect the "actual" safety invariants. I agree that the stuff about atomic/non-atomic accesses being synchronized is tricky to explain and easy to mess up, but if you are using atomics, chances are, you already doing things that are hard to explain and easy to mess up 🙃 

Either way, I think that leaving what is currently written as something like "the «actual» conditions are trivially satisfied if ..." may be a good idea. Like we could have both the easy to understand, but a bit too restrictive _and_ the actually correct, but complex rules in the docs.

**Comment by tgross35** [comment]
@WaffleLapkin I updated the docs to say the following:

```diff
+    /// * The value behind `ptr` must not be accessed through non-atomic operations whenever
+    ///   the `AtomicBool` is used in a `Sync` context (e.g. read or modified in a different
+    ///   thread).
```

does that sound better?

Sorry @GuillaumeGomez  I have absolutely no clue how your commit wound up in this PR

**Comment by WaffleLapkin** [comment]
@tgross35 I think it's might be a bit unclear and we might want to provide the very explicit "hey, you want a happens-before relationship". But yes it's better in my opinion and we might tweak docs later.

**Comment by WaffleLapkin** [comment]
I had something like this in mind:

> * Non-atomic accesses to the value behind `ptr` must have a happens-before relationship with atomic accesses to the returned atomic or vice-versa
>   * i.e. you need to synchronize the periods of time when you access the value atomically and non-atomically
>   * Note that this requirement is trivially satisfied if you do not access the value non-atomically for the duration of lifetime `'a`
>   * This requirement is also trivially satisfied if you do all accesses (atomic or not) on the same thread

But you may want to ask people who know C++ atomics better (maybe ask on zulip for suggestions for this doc?).

**Comment by Zenithsiz** [comment]
I believe at some point during this feature's lifetime there was a discussion on whether mixed loads/stores were allowed.

Since this is being added, was there a concensus towards that?

For example, would the following be allowed now?

```rust
// Assume correct alignment for `AtomicU32`
let mut value = 0_u32;

let value_ptr = &mut value as *mut u32;
let ptr = AtomicU32::from_ptr(value_ptr);
let ptr_lo = AtomicU16::from_ptr(value_ptr.cast::<u16>());
let ptr_hi = AtomicU16::from_ptr(value_ptr.cast::<u16>().add(1));

ptr.store(0x00020001, Ordering::Release);
let lo = ptr_lo.load(Ordering::Acquire);
let hi = ptr_hi.load(Ordering::Acquire);

assert_eq!(lo, 0x0001);
assert_eq!(hi, 0x0002);
```

For reference, [here](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=15607e48e6795ec28648dea61d59ce6a) is a playground, and miri currently doesn't complain.

**Comment by tgross35** [comment]
That is an interesting and somewhat spooky, any chance you have a link to the discussion you are referencing? I don't see it on the related issues.

I am far from an expert here but I think it is fine as-is, since `ptr` and `lo` have the same memory addresses it couldn't swap those, and it can't swap the acquires. I think that if you rearranged it to `ptr.store, lo.load, hi.load` then maybe it would be allowed to do something like this since `ptr` and `lo` don't have the same memory address:

```rust
let hi = ptr_hi.load(Ordering::Acquire);
ptr.store(0x00020001, Ordering::Release);
let lo = ptr_lo.load(Ordering::Acquire);
```

But maybe it checks the sizes of the value and not just the address to decide what's related. In any case I feel like we probably don't need to say anything specific here, this seems like pretty fragile behavior

I think @m-ou-se literally wrote the book on atomics and might know more. cc @RalfJung too for the miri question

**Comment by RalfJung** [comment]
We'd like to detect this in Miri but it hasn't been implemented yet: https://github.com/rust-lang/miri/issues/2303.

There's also the question about racing atomic and non-atomic accesses, which are not possible in C++ so given that we say "we are using the C++ memory model" it is unclear whether we can allow them.

**Comment by Zenithsiz** [comment]
@tgross35 I found it, it was in [this issue](https://github.com/rust-lang/rust/issues/76314#issuecomment-887112198).

There didn't seem to be much of a discussion after it being mentioned, however.

**Comment by tgross35** [comment]
Thanks for the link, I guess we can probably safely say it’s out of scope. I think that for it to work, it would need to treat every address within a parent atomic the same when ordering is evaluated - but I have a feeling that if the c++ model doesn’t support it, then probably only the base address is being checked.

Not to say that somebody couldn’t roll their own, but I assume those use cases are probably already doing what they need through casting

**Comment by RalfJung** [comment]
I think the docs for these functions should explicitly say that mixed-sized atomic accesses are forbidden.

**Comment by dtolnay** [comment]
@rustbot author
to incorporate any takeaways from the above discussion into the docs.

**Comment by tgross35** [comment]
I took @WaffleLapkin's comment with some (I think) simplified wording, and added a note about overlapping atomics

@rustbot ready

**Review by dtolnay — APPROVED** [review]
Great! Thank you!

**Comment by RalfJung** [comment]
> I think the docs for these functions should explicitly say that mixed-sized atomic accesses are forbidden.

I don't think this has been done? @tgross35 could you file a follow-up PR fixing the docs? This is important, we don't want people to assume that these are implicitly allowed.

File: library/core/src/sync/atomic.rs [hunk 1]
```diff
@@ -346,13 +346,21 @@ impl AtomicBool {
...
     /// * `ptr` must be [valid] for both reads and writes for the whole lifetime `'a`.
-    /// * The value behind `ptr` must not be accessed through non-atomic operations for the whole lifetime `'a`.
+    /// * Non-atomic accesses to the value behind `ptr` must have a happens-before relationship
+    ///   with atomic accesses via the returned value (or vice-versa).
+    ///   * In other words, time periods where the value is accessed atomically may not overlap
+    ///     with periods where the value is accessed non-atomically.
+    ///   * This requirement is trivially satisfied if `ptr` is never used non-atomically for the
+    ///     duration of lifetime `'a`. Most use cases should be able to follow this guideline.
+    ///   * This requirement is also trivially satisfied if all accesses (atomic or not) are done
+    ///     from the same thread.
```
Comment (RalfJung) [inline library/core/src/sync/atomic.rs:359]: Specifically I was looking for the overlapping atomics part here and I am not seeing it.

File: library/core/src/sync/atomic.rs [hunk 2]
```diff
@@ -2111,14 +2129,25 @@ macro_rules! atomic_int {
...
+            /// * Non-atomic accesses to the value behind `ptr` must have a happens-before
+            ///   relationship with atomic accesses via the returned value (or vice-versa).
+            ///   * In other words, time periods where the value is accessed atomically may not
+            ///     overlap with periods where the value is accessed non-atomically.
+            ///   * This requirement is trivially satisfied if `ptr` is never used non-atomically
+            ///     for the duration of lifetime `'a`. Most use cases should be able to follow
+            ///     this guideline.
+            ///   * This requirement is also trivially satisfied if all accesses (atomic or not) are
+            ///     done from the same thread.
+            /// * This method should not be used to create overlapping or mixed-size atomic
```
Comment (RalfJung) [inline library/core/src/sync/atomic.rs:2145]: This should say "must not".

Looks like some methods got this note and some didn't; it should be present on all of these methods.

**Comment by tgross35** [comment]
@RalfJung I will follow up in https://github.com/rust-lang/rust/pull/116762

**Comment by RalfJung** [comment]
@taiki-e [points out](https://github.com/rust-lang/miri/issues/2303#issuecomment-1774035373) an interaction between the caveats we are adding here to stay within the bounds of the C++ model, and [this comment](https://github.com/rust-lang/rust/pull/115577#issuecomment-1710786587) by @dtolnay:

> As I understand it, the new guarantee introduced here would make the following sound: we want a value that can be atomically loaded but the backing data might either be the read-write memory where interior mutable values ordinarily go, or could be readonly memory on certain architectures that have this new guarantee.

Under the constraints imposed by this PR, I think that code would *not* be sound. And indeed there is no way to write such code soundly in C++, so we should be extremely careful with allowing it in Rust.

**Pull request merged**